### PR TITLE
Rework database initialization

### PR DESCRIPTION
**Pull request by araspitzu**
This PR aims at separating the code that initializes the database from the NodeParams which is the application-level object carrying configuration information. Now test can be executed providing a complete stub interface (not included in this PR) and StartupSpec can run without creating files during its execution.

### CODE REVIEW

File: eclair-core/src/main/scala/fr/acinq/eclair/NodeParams.scala [hunk 1]
```diff
@@ -129,28 +124,31 @@ object NodeParams {
...
 
-  def makeNodeParams(datadir: File, config: Config, keyManager: KeyManager, torAddress_opt: Option[NodeAddress]): NodeParams = {
+  def loadDb(chaindir: File): AbstractDb = {
+    chaindir.mkdir()
+    val sqliteEclair = DriverManager.getConnection(s"jdbc:sqlite:${new File(chaindir, "eclair.sqlite")}")
+    val sqliteNetwork = DriverManager.getConnection(s"jdbc:sqlite:${new File(chaindir, "network.sqlite")}")
+    val sqliteAudit = DriverManager.getConnection(s"jdbc:sqlite:${new File(chaindir, "audit.sqlite")}")
+    SqliteUtils.obtainExclusiveLock(sqliteEclair) // there should only be one process writing to this file
+
+    val db = new {} with AbstractDb {
```
[the remark this answers is not in the document]
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/NodeParams.scala]: Done in bd12bd1

File: eclair-core/src/main/scala/fr/acinq/eclair/db/AbstractDb.scala [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package fr.acinq.eclair.db
+
+trait AbstractDb {
```
[the remark this answers is not in the document]
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/db/AbstractDb.scala]: Done in bd12bd1

File: eclair-core/src/test/scala/fr/acinq/eclair/TestConstants.scala [hunk 3]
```diff
@@ -181,4 +170,13 @@ object TestConstants {
...
+  def inMemoryDb(connection: Connection = sqlite()): AbstractDb = new {} with AbstractDb {
+    override def network(): NetworkDb = new SqliteNetworkDb(connection)
+    override def audit(): AuditDb = new SqliteAuditDb(connection)
+    override def channels(): ChannelsDb = new SqliteChannelsDb(connection)
+    override def peers(): PeersDb = new SqlitePeersDb(connection)
+    override def payments(): PaymentsDb = new SqlitePaymentsDb(connection)
+    override def pendingRelay(): PendingRelayDb = new SqlitePendingRelayDb(connection)
+  }
+
 }
```
[the remark this answers is not in the document]
Comment (araspitzu) [reply eclair-core/src/test/scala/fr/acinq/eclair/TestConstants.scala]: Yes and the `def` was definitely a mistake! I've addressed them in bd12bd1 and  7c4bb06

File: eclair-core/src/main/scala/fr/acinq/eclair/NodeParams.scala [hunk 4]
```diff
@@ -129,28 +124,31 @@ object NodeParams {
     }
   }
 
-  def makeNodeParams(datadir: File, config: Config, keyManager: KeyManager, torAddress_opt: Option[NodeAddress]): NodeParams = {
+  def loadDb(chaindir: File): AbstractDb = {
```
[the remark this answers is not in the document]
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/NodeParams.scala]: Done in 0aa5385

File: eclair-core/src/main/scala/fr/acinq/eclair/NodeParams.scala [hunk 5]
```diff
@@ -129,28 +124,31 @@ object NodeParams {
...
+    val db = new {} with AbstractDb {
+      override def network(): NetworkDb = new SqliteNetworkDb(sqliteNetwork)
+      override def audit(): AuditDb = new SqliteAuditDb(sqliteAudit)
+      override def channels(): ChannelsDb = new SqliteChannelsDb(sqliteEclair)
+      override def peers(): PeersDb = new SqlitePeersDb(sqliteEclair)
+      override def payments(): PaymentsDb = new SqlitePaymentsDb(sqliteEclair)
+      override def pendingRelay(): PendingRelayDb = new SqlitePendingRelayDb(sqliteEclair)
+    }
+    db
+  }
```
[the remark this answers is not in the document]
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/NodeParams.scala]: Done in 7c4bb06

File: eclair-core/src/main/scala/fr/acinq/eclair/NodeParams.scala [hunk 6]
```diff
@@ -129,28 +124,31 @@ object NodeParams {
     }
   }
 
-  def makeNodeParams(datadir: File, config: Config, keyManager: KeyManager, torAddress_opt: Option[NodeAddress]): NodeParams = {
+  def loadDb(chaindir: File): AbstractDb = {
+    chaindir.mkdir()
```
[the remark this answers is not in the document]
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/NodeParams.scala]: The problem is that we need to read the config to know which the folder of the db, **chaindir** is extracted from the config. By passing a function we can have both, the name of the folder is read from the config (see "chain") but we can still occasionally mock it.

File: eclair-core/src/main/scala/fr/acinq/eclair/db/Databases.scala [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+object Databases {
+
+  /**
+    * Given a parent folder it creates or loads all the databases from a JDBC connection
+    * @param chaindir
+    * @return
+    */
+  def createOrLoadSQLiteWithJDBC(chaindir: File): Databases = {
```
[the remark this answers is not in the document]
Comment (sstone) [reply eclair-core/src/main/scala/fr/acinq/eclair/db/Databases.scala]: I think that name should explicitly reference JDBC , maybe 
```suggestion
  def sqliteJDBC(chaindir: File): Databases = {
```

[the remark this answers is not in the document]
Comment (araspitzu) [reply eclair-core/src/main/scala/fr/acinq/eclair/db/Databases.scala]: Agree but then let's rename [this](https://github.com/ACINQ/eclair/pull/911/files#diff-77701f4858621ac99b887487f9bf95adR42) too as "sqliteInMemory()"

**Pull request merged**